### DISCUSSION TITLE
Components break if interaction occurs while browser is offline

### DISCUSSION THREAD

**Discussion by joseph-d** [livewire/livewire#6957] [7. Bugs (archived)] [answered]
### Livewire version

3 (works fine on 2, haven't tested 1)

### Which PHP version are you using?

PHP 8.2

### Steps To Reproduce

**Steps to reproduce the expected error that occurs when interacting with a Livewire component from a recently-offline browser:**

1. Load any page that is a Livewire full page component or has Livewire components within it. A quick and easy public example would be the Filament customer list page at https://demo.filamentphp.com/shop/customers but literally any other Livewire page will be suitable too.
2. Click the "Edit" button next to a customer until you find one with at least 2 attached addresses which we will use to illustrate the issue
3. Click "Detach" on one of the addresses and confirm the dialog that appears
4. Now, take the browser offline. This could be done by putting a mobile in flight mode or using Chrome dev tools Network tab to choose the "Offline" state
5. Try clicking a "Detach" link again. The spinner will start spinning indefinitely
7. In Chrome devtools the network request will show as "(failed)" 
8. The console log will show 2 errors: a POST error (`net::ERR_INTERNET_DISCONNECTED`) and `Uncaught (in promise) TypeError: Failed to fetch`

So far, _all of this is perfectly normal_ and exactly what I would expect to happen.

**Steps to illustrate what happens after the connection is restored:**

1. Restore the connection by disabling flight mode or change the Chrome devtools network status back to the default "No throttling" setting to restore the normal network connection.
2. Observe that the spinner is still spinning next to the last "Detach" button you pressed in Filament
3. Try to click that Detach button, another Detach button or, indeed, any other `wire:click`. 
4. Observe that the mouse cursor does not change to a pointer and clicking `wire:click` links does not cause any reaction in either the Chrome devtools network tab nor does it create an entry into the console log.
5. The Livewire component is now effectively "broken" from the point of view of an end user and the only way to fix it is to refresh the page. But there is no way to prompt the user into doing this, meaning that non-technical users will simply think that it has "stopped working" because everything else on their device is working properly.

**Why I believe this is a significant issue:**

While most broadband / wifi connections are stable enough for this to go completely unnoticed, it's a lot more common for mobile browsers to flip between online and offline when a user is travelling around. Users know that this happens but they are also used to being notified about this via an alert telling them that they're offline (JS frameworks) or the familiar `No Internet: ERR_INTERNET_DISCONNECTED` that comes from Chrome and other browsers when clicking on any link when a browser is offline.

However, in the case described above, there is zero feedback to the user. The only thing we can do is use `wire:offline` to display a message to the user when the browser is offline. But in the example above, we're talking about a scenario where the browser is now _online_ again which means that the `wire:offline` message would not be showing anymore.

As interacting with a component broken in this manner doesn't produce a network error or anything in the console log, I am unsure about what to do in terms of identifying the cause. I am also at a loss to know where to start at creating a workaround.

**How I believe this could be remedied**

Well, I have no idea actually, but as the issue didn't occur in Livewire 2 and the error showing in Livewire 3 is a type error, hopefully someone who is good with Javascript can easily identify the problem and fix it.

### Additional information

This issue does not occur in Livewire 2. The console shows `net::ERR_INTERNET_DISCONNECTED` in both v2 and v3, but in v2 there is no  `Uncaught (in promise) TypeError: Failed to fetch` error. 

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by joseph-d** [comment]
Perhaps my initial explanation was a little bit too long-winded. Here's a simpler example along with a video:

**Code using Volt (although issue is not Volt-specific)**

```php
<?php
 
use Livewire\Volt\Component;
 
new class extends Component {
    public $count = 0;
 
    public function increment()
    {
        $this->count++;
    }
} ?>
 
<div>
    <h1>{{ $count }}</h1>
    <button wire:click="increment">+</button>
</div>
```

**Video showing the following:**

1. Component works initially
2. Component stops working when no network
3. Component no longer functions when network restored
4. Component functions after whole-page refresh

[Peek 2023-10-23 09-55.webm](https://github.com/livewire/livewire/assets/1070155/11867abe-b893-4c11-84cc-4e1657e94ffa)

**Reply by timetorock** [reply]
Did you figured that out how to fix this? Faced same behaviour. If browser goes offline and user click something - component broke until page reload.

**Reply by joseph-d** [reply]
No, I haven't figured out how to fix it unfortunately. It's really a shame because in some situations it's stopped me from being able to use Livewire.

I've found it difficult to get anyone interested in this as an issue, despite it being seemingly unique to Livewire. Vue and even HTMX don't seem to suffer from the same problem, so I'm convinced it's something that's fixable but unfortunately I wouldn't personally know where to start.

**Reply by timetorock** [reply]
Currently, I've resolved this issue with wire:offline.attr="disabled": https://livewire.laravel.com/docs/offline. It blocks buttons while offline and prevent this error. I agree it's not ideal, but at least, it is working.

**Reply by JustGAST** [reply]
Thank you for workaround! Unfortunately it does not help if connection were lost during the request. After returning online everything is unresponsive.

**Reply by JustGAST** [reply]
My workaround is to save data to local storage on every model interaction, reload page when returned online and restore state from local storage.

Reload page on returning online:
```
window.addEventListener('online', () => {
    window.location.reload();
});
```

UPD: Forgot to mention that I was forced to transfer state from livewire component to alpine x-data, and entangle livewire attribute. That allowed me to work with local storage.

**Reply by joseph-d** [reply]
Yes, that's right. For example if the user was on a train with a spotty off/on connection they could press the button at precisely the worng time and then from that moment on they will think that the website is broken because everything else on their phone will work except our livewire app, and they won't know that they have to refresh it.

**Comment by JustGAST** [comment]
Same issue! Did you open github issue?
Same problem here: https://laracasts.com/discuss/channels/livewire/liverwire?reply=868260

**Reply by joseph-d** [reply]
No, I didn't open an issue. I wanted it get other people's opinion with a discussion here first. See if there are any other ways to look at it.

**Reply by shubdragon** [reply]
i checked my project also have same issue the loader on page load infinitely does and setting no throttling and click on any new livewire action then nothing respond to click and cursor no longer  change to pointer  on relevant elements

**Comment by ArtMin96** [comment]
Same issue here

**Comment by rstanislav** [comment]
In v2 i'm using this at the end of app.blade.php:

```
window.livewire.onError((statusCode) => {
        console.log('livewire error - auto page reload');
        window.location.reload();
        return false;
    });
```

In real big livewire project there is always be errors, and after any of it it becomes unresponsive, code above resolves it and user 99.9% of time didnt even notice reload.

**Reply by JustGAST** [reply]
Looks like there is no `onError` method in Livewire 3, and version 2 has it, according to documentation:
https://laravel-livewire.com/docs/2.x/reference#global-livewire-js

On version 3
```
Console log:
Livewire.onError
undefined
```

**Comment by JustGAST** [comment]
I believe there are enough information to open issue. @joseph-d made a great job providing problem demo. I'll open an issue.

**Comment by JustGAST** [comment]
This logs to console on request made offline. Maybe it'll be helpful.

```
POST http://127.0.0.1:8000/livewire/update net::ERR_INTERNET_DISCONNECTED
(anonymous) @ livewire.js?id=2f6e5d4d:534
queueNewRequestAttemptsWhile @ livewire.js?id=2f6e5d4d:612
sendRequestToServer @ livewire.js?id=2f6e5d4d:506
(anonymous) @ livewire.js?id=2f6e5d4d:492
(anonymous) @ livewire.js?id=2f6e5d4d:500
setTimeout (async)
bundleMultipleRequestsTogetherIfTheyHappenWithinFiveMsOfEachOther @ livewire.js?id=2f6e5d4d:499
triggerSend @ livewire.js?id=2f6e5d4d:491
(anonymous) @ livewire.js?id=2f6e5d4d:649
(anonymous) @ livewire.js?id=2f6e5d4d:606
waitUntilTheCurrentRequestIsFinished @ livewire.js?id=2f6e5d4d:602
requestCall @ livewire.js?id=2f6e5d4d:647
(anonymous) @ livewire.js?id=2f6e5d4d:4174
runIfTypeOfFunction @ livewire.js?id=2f6e5d4d:1317
(anonymous) @ livewire.js?id=2f6e5d4d:1305
tryCatch @ livewire.js?id=2f6e5d4d:1223
evaluate @ livewire.js?id=2f6e5d4d:1247
(anonymous) @ livewire.js?id=2f6e5d4d:7143
callAndClearComponentDebounces @ livewire.js?id=2f6e5d4d:7128
execute @ livewire.js?id=2f6e5d4d:7142
x-on:click @ livewire.js?id=2f6e5d4d:7151
(anonymous) @ livewire.js?id=2f6e5d4d:1267
tryCatch @ livewire.js?id=2f6e5d4d:1223
(anonymous) @ livewire.js?id=2f6e5d4d:3751
handler4 @ livewire.js?id=2f6e5d4d:3123
(anonymous) @ livewire.js?id=2f6e5d4d:3188
(anonymous) @ livewire.js?id=2f6e5d4d:3125
livewire.js?id=2f6e5d4d:534 
        
        
       Uncaught (in promise) TypeError: Failed to fetch
    at livewire.js?id=2f6e5d4d:534:28
    at queueNewRequestAttemptsWhile (livewire.js?id=2f6e5d4d:612:11)
    at sendRequestToServer (livewire.js?id=2f6e5d4d:506:11)
    at livewire.js?id=2f6e5d4d:492:7
    at livewire.js?id=2f6e5d4d:500:7
(anonymous) @ livewire.js?id=2f6e5d4d:534
queueNewRequestAttemptsWhile @ livewire.js?id=2f6e5d4d:612
sendRequestToServer @ livewire.js?id=2f6e5d4d:506
(anonymous) @ livewire.js?id=2f6e5d4d:492
(anonymous) @ livewire.js?id=2f6e5d4d:500
await in (anonymous) (async)
(anonymous) @ livewire.js?id=2f6e5d4d:492
(anonymous) @ livewire.js?id=2f6e5d4d:500
setTimeout (async)
bundleMultipleRequestsTogetherIfTheyHappenWithinFiveMsOfEachOther @ livewire.js?id=2f6e5d4d:499
triggerSend @ livewire.js?id=2f6e5d4d:491
(anonymous) @ livewire.js?id=2f6e5d4d:649
(anonymous) @ livewire.js?id=2f6e5d4d:606
waitUntilTheCurrentRequestIsFinished @ livewire.js?id=2f6e5d4d:602
requestCall @ livewire.js?id=2f6e5d4d:647
(anonymous) @ livewire.js?id=2f6e5d4d:4174
runIfTypeOfFunction @ livewire.js?id=2f6e5d4d:1317
(anonymous) @ livewire.js?id=2f6e5d4d:1305
tryCatch @ livewire.js?id=2f6e5d4d:1223
evaluate @ livewire.js?id=2f6e5d4d:1247
(anonymous) @ livewire.js?id=2f6e5d4d:7143
callAndClearComponentDebounces @ livewire.js?id=2f6e5d4d:7128
execute @ livewire.js?id=2f6e5d4d:7142
x-on:click @ livewire.js?id=2f6e5d4d:7151
(anonymous) @ livewire.js?id=2f6e5d4d:1267
tryCatch @ livewire.js?id=2f6e5d4d:1223
(anonymous) @ livewire.js?id=2f6e5d4d:3751
handler4 @ livewire.js?id=2f6e5d4d:3123
(anonymous) @ livewire.js?id=2f6e5d4d:3188
(anonymous) @ livewire.js?id=2f6e5d4d:3125
```

**Comment by JustGAST** [comment]
Unfortunately, commit and request hooks also not working offline:
https://livewire.laravel.com/docs/javascript#request-hooks

**Comment by timetorock** [comment]
The problem is that Livewire do not handle errors when request failed in the middle. For example because user lost internet connection. So it can't be catch by any livewire hook, etc, as it's internal error. After this -> Livewire freeze and nothing work.

More details here:

https://github.com/livewire/livewire/discussions/7472

**Comment by rstanislav** [comment]
Is there any updates on this issue? This is huge problem, we are forced to use v2 only because of this bug =(

**Reply by JustGAST** [reply]
Does not v2 have this bug?

**Reply by joseph-d** [reply]
Sorry, this was my mistake. Actually, it works fine in Livewire 2 and this appears to be a new issue which happens in Livewire 3. I have updated the original post to reflect this.

**Comment by sebtrucker** [comment]
Any updates about this issue ? The only solution I found so far it's to refresh the page everytime the connection is lost but it's really not practical :/

**Reply by danie-ramdhani** [reply, marked as answer]
i think it's fixed in v.3.4.7 https://github.com/livewire/livewire/pull/7972

**Reply by sebtrucker** [reply]
I was on the 3.4.6 so I made the update and it's working ! thanks !
